### PR TITLE
Restrict loading to root or below.

### LINKED ISSUE #693: File paths may not step up.
Currently, all file paths mentioned in a kustomization file must be defined relative to the location of the kustomization file.

This allows (e.g.) the following fishing attack.  Someone writes a kustomization with a configmap generator that reads data from `../../../etc/passwd`, and sends it to a victim, hopes the victim performs the kustomization in the correct working directory, and hopes to subsequently look at the cluster to see the data.

To prevent this, the following restriction is added:

All files paths mentioned in a kustomization file (resources, patches, config data) must refer to files in or below `.`, the location of the kustomization file.

Bases are not subject to this, and may be specified as living in sibling directories (or in remote URLs).  A base can be viewed as an “import”, and its data files are subject to the same restriction but relative to that base.

Symlinks will be purged (e.g. via `filepath.EvalSymlinks`).

### PR DESCRIPTION
**Pull request by monopole**
Kustomization files refer to two kinds of files outside the kustomization file itself:

 - _bases_ (other kustomizations)
   These can be git URLs, or paths specified relative to the __root__ (as long as they don't go strictly "up").

 - _data paths_  for the current kustomization
   Things like resources, patches, data for configmaps, etc.
   These must be paths specified relative to the __root__

where __root__ is the directory containing the kustomization file specifying these things.

This PR imposes an additional requirement on _data paths_. 

They must still be __root__ relative paths, but additionally they cannot go up out of __root__, either explicitly, or by following symbolic links.  They can only refer to files in or below __root__.

This will require an increment in semver major, since it disallows certain file loading patterns.

No existing examples or documentation need to change, since this file arrangement wasn't used.

Example of disallowed behavior and how to fix:

Supposed a single data file `foo/common/data.txt` is referenced in two similar configmap generator rules in two different directories `foo/overlay1` and `foo/overlay2`.  This is now disallowed, because the data is outside the two respective kustomization roots.

The fix is to simply add a kustomization file to the directory containing the data, and refer to that as a base from the kustomizations in the two overlays (i.e. refer to the base, not the data).  It remains OK to go up and over when referring to a base (though not directly up, since that can result in cycles).

A followup PR should include an example discussing/defining this.

Fixes #693

### CODE REVIEW

File: pkg/loader/fileloader.go [hunk 1]
```diff
@@ -205,13 +217,42 @@ func isPathEqualToOrAbove(path string, roots []string) error {
...
 
-// Load returns content of file at the given relative path.
+// Load returns content of file at the given relative path,
+// else an error.  The path must refer to a file in or
+// below the current Root().
 func (l *fileLoader) Load(path string) ([]byte, error) {
 	if filepath.IsAbs(path) {
-		return nil, fmt.Errorf(
-			"must use relative path; '%s' is absolute", path)
+		return nil, l.loadOutOfBounds(path)
```
Comment (Liujingfang1) [inline pkg/loader/fileloader.go]: The previous error message looks better. We can leave it unchanged.

Comment (monopole) [reply pkg/loader/fileloader.go]: Message updated.

This is now a change from
```
must use relative path; '/etc/tls.key' is absolute.
```
to
```
"security; file `/etc/tls.key` is not in or below `/the/kustomization/root`,
```
How's that?

File: pkg/loader/fileloader.go [hunk 2]
```diff
@@ -205,13 +217,42 @@ func isPathEqualToOrAbove(path string, roots []string) error {
...
+// below Root() from purely a path perspective (no
+// check for actual file existence). For this to work,
+// both the given argument "path" and l.Root() must
+// be cleaned, absolute paths, and l.Root() must be
+// a directory.  Both conditions enforced elsewhere.
+func (l *fileLoader) isInOrBelowRoot(path string) bool {
+	if l.Root() == string(filepath.Separator) {
+		return true
+	}
+	return strings.HasPrefix(
```
[the remark this answers is not in the document]
Comment (monopole) [reply pkg/loader/fileloader.go]: Thanks much for those links.

I've updated this comment and the overall class doc.

**Comment by DirectXMan12** [comment]
new comments look good
/lgtm

**Pull request merged**